### PR TITLE
Fallback to http proxy when https proxy is invalid

### PR DESCRIPTION
**Pull request by UsmannK**
Currently if https_proxy is not set then we fallback to http_proxy. This PR adds functionality to fallback to http_proxy in the case that https_proxy is set but is invalid. Currently, we would just fail to no proxy in that case.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
CLA GitHub bot_

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 1]
```diff
@@ -81,12 +73,31 @@ static char* get_http_proxy_server(char** user_cred) {
     proxy_name = nullptr;
   }
   gpr_free(authority_strs);
-done:
-  gpr_free(uri_str);
   grpc_uri_destroy(uri);
   return proxy_name;
 }
 
+grpc_uri* get_proxy_uri(char* proxy_env_var) {
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/http_proxy.cc]: Can we pass the input as `const char*`?

Suggest naming the parameter as `proxy_env_var_name` or `proxy_env_var_key`.

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 2]
```diff
@@ -50,18 +50,10 @@ static char* get_http_proxy_server(char** user_cred) {
   /* Prefer using 'https_proxy'. Fallback on 'http_proxy' if it is not set. The
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/http_proxy.cc]: Please update this comment with the new behavior.

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 3]
```diff
@@ -81,12 +73,31 @@ static char* get_http_proxy_server(char** user_cred) {
...
   grpc_uri_destroy(uri);
   return proxy_name;
 }
 
+grpc_uri* get_proxy_uri(char* proxy_env_var) {
+  char* uri_str = gpr_getenv(proxy_env_var); 
+  if (uri_str == nullptr) {
+    return nullptr;
+  }
+
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/http_proxy.cc]: Please remove blank lines in functions.

File: src/core/ext/filters/client_channel/http_proxy.cc [hunk 4]
```diff
@@ -81,12 +73,31 @@ static char* get_http_proxy_server(char** user_cred) {
...
   }
   gpr_free(authority_strs);
-done:
-  gpr_free(uri_str);
   grpc_uri_destroy(uri);
   return proxy_name;
 }
 
+grpc_uri* get_proxy_uri(char* proxy_env_var) {
+  char* uri_str = gpr_getenv(proxy_env_var);
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/http_proxy.cc]: Please use `const char*` if appropriate.

**Comment by markdroth** [comment]
I'm not sure whether or not we actually want to do this.  I think Yash should look at this when he gets back.

**Comment by UsmannK** [comment]
> I'm not sure whether or not we actually want to do this. I think Yash should look at this when he gets back.

Could you explain why not? My thinking is along the lines of: https_proxy and http_proxy are generally set to the same thing. When they aren't, it's likely because https_proxy has an https uri scheme (meaning it is using TLS to not only connect from the proxy to the destination but also from the host to the proxy). GRPC doesn't support that config, so IMO we should fall back to http_proxy just like we already do when there is no https_proxy set. I don't see why failing straight to no proxy at all is a desirable solution.

**Comment by UsmannK** [comment]
Gotcha, was just wondering if there is a specific concern.

**Comment by yashykt** [comment]
Please refer to the discussion in #9561 and continue the discussion there if needed. In short, we do want to use http_proxy for CONNECT. Currently, it is set to be a backup if https_proxy was not set at all, and we might even remove that in upcoming versions if we determine that it would be better that way.

**Comment by UsmannK** [comment]
@yashykt This is not the same issue. This refers to the fact that grpc does not support the https URI for the host<->proxy communication and falls back to no proxy, instead of http_proxy. I don't see any discussion there as to the reasoning behind having inconsistent fallback behavior.

How grpc acts today: 
1. I set `export https_proxy=https://proxy.mycorp.co`
2. grpc sees this invalid uri and falls straight back to not using any proxy at all

After this change:
1. I set `export https_proxy=https://proxy.mycorp.co`
2. grpc sees this invalid uri and falls back to using http_proxy
3. grpc validates `http_proxy`
4. Either use `http_proxy` if valid or use no proxy otherwise.

**Comment by yashykt** [comment]
@UsmannK You are right in that gRPC does not use the fallback if https_proxy is set but is invalid. But this in itself is not inconsistent. I would prefer to remove the fallback altogether as the the discussion in #9561 explains.

**Comment by AspirinSJL** [comment]
According to @yashykt 's comment above, I am going to close this PR.

**Pull request closed without merging**